File: 0547_findCircleNum.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        result = 0
        visited = set()
        for i in range(n):
            if i not in visited:
                visited.add(i)
                temp = [i]
                self.dfs(isConnected, n, i, visited, temp)
                result += 1
        return result
        
    def dfs(self, isConnected, n, i, visited, temp):
        for j in range(n):
            if isConnected[i][j] == 1 and j not in visited:
                visited.add(j)
                temp.append(j)
                self.dfs(isConnected, n, j, visited, temp)
                
# Alternative solution

class UnionFind:
    def __init__(self, n):
        self.parent = [i for i in range(n)]

    def find(self, a):
        while a != self.parent[a]:
            a = self.parent[a]
        
        return a

    def union(self, a, b):
        root_a = self.find(a)
        root_b = self.find(b)

        if root_a != root_b:
            self.parent[root_b] = root_a

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        uf = UnionFind(n)

        for i in range(n):
            for j in range(i + 1, n):
                if isConnected[i][j]:
                    uf.union(i, j)
        
        for i in range(n):
            uf.parent[i] = uf.find(uf.parent[i])

        return len(set(uf.parent))

File: 0501-1000/0547_findCircleNum.java
class UnionFind {
    int[] parent;
    int[] rank;

    public UnionFind(int n) {
        this.parent = new int[n];
        this.rank = new int[n];

        for (int i = 0; i < n; i++) {
            this.parent[i] = i;
        }
    }

    public int find(int a) {
        while (a != this.parent[a]) {
            a = this.parent[a];
        }

        return a;
    }

    public void union(int a, int b) {
        int aRoot = find(a);
        int bRoot = find(b);

        if (aRoot != bRoot) {
            if (this.rank[bRoot] < this.rank[aRoot]) {
                this.parent[bRoot] = aRoot;
            } else if (this.rank[bRoot] == this.rank[aRoot]) {
                this.parent[bRoot] = aRoot;
                this.rank[aRoot]++;
            } else {
                this.parent[aRoot] = bRoot;
            }
        }
    }
}

class Solution {
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        UnionFind unionFind = new UnionFind(n);
        HashSet<Integer> parentSet = new HashSet<>();

        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (isConnected[i][j] == 1) {
                     unionFind.union(i, j);
                }
            }
        }

        for (int i = 0; i < n; i++) {
            parentSet.add(unionFind.find(i));
        }

        return parentSet.size();
    }
}
